CHANGELOG.md

```diff
@@ -81,6 +81,7 @@ By @wumpf in [#6069](https://github.com/gfx-rs/wgpu/pull/6069), [#6099](https://
 
 - Reduce the amount of debug and trace logs emitted by wgpu-core and wgpu-hal. By @nical in [#6065](https://github.com/gfx-rs/wgpu/issues/6065)
 - `Rg11b10Float` is renamed to `Rg11b10UFloat`. By @sagudev in [#6108](https://github.com/gfx-rs/wgpu/pull/6108)
+- Removed `.global_id()` getters from wgpu objects. By @teoxoy [#6134](https://github.com/gfx-rs/wgpu/pull/6134)
```

Comment:
We can't just remove this - it's used externally for caching and stuff. We should generate ids like we do for wasm. Longer term this problem will be solved by our types implementing partialeq/hash

Reply:
If we now have to generate IDs for both backends, wouldn't it make sense to have users interested in this functionality do this?

I looked at the first 3 pages of dependents on crates io and none of them use the `global_id`.

wgpu types don't implement Clone either, they are unique and so was their ID, so I'm not sure how the ID would be useful for caching.

Looping in @TheOnlyMrCat who was the author of https://github.com/gfx-rs/wgpu/pull/3104.

Reply:
> If we now have to generate IDs for both backends, wouldn't it make sense to have users interested in this functionality do this?

The problem is that they may not control all creation and deletion code (think middleware). This gives a pretty easy way to determine a wgpu resouece's identity.

Reply:
I see, what do you think about using the pointers of the internal Boxes to implement `PartialEq`, `Eq`, and `Hash` for the resource types directly? That would also solve https://github.com/gfx-rs/wgpu/issues/6080.

Reply:
Sorry got called to an appointment in between responses 😛

Yeah that would work, we would need to make sure that all resources had those impls.

Reply:
While we are at it, I feel like we should reconsider some of the objects implementing this and do it only for those that have use cases.

My proposal:

Yes:

- Buffer
- Texture
- TextureView
- Sampler
- QuerySet

Maybe?:

- BindGroupLayout (this won't be equal with others but will be compatible)
- BindGroup
- ShaderModule
- PipelineLayout
- RenderPipeline
- ComputePipeline
- RenderBundle

No:

- Surface
- Adapter
- Device
- Queue

Thoughts on this? cc @dfellis

Reply:
I agree with your Yes and No.

For the Maybe, the `RenderPipeline` and `ComputePipeline` immediately feel like "yes" to me: They could be useful as the keys to a `HashMap` that caches a `HashSet` of `BindGroup`s that have already been constructed for that operation, letting a bunch of setup work get short-circuited if this particular shader on this particular collection of Buffers/Textures/etc has been executed before. (This reasoning immediately marks `BindGroup` as "yes," as well.)

`BindGroupLayout` is also "yes," I think, because you can share the same bind group layout between pipelines on purpose. It could be useful to check different pipelines' `get_bind_group_layout` return values actually being the same underlying object so you know whether or not you can reuse bind groups between them.

I'm not so familiar with `RenderBundle` (it looks like some sort of hardware-specific optimization for the command encoder?) so I don't know if people would want to cache multiple kinds of these bundles and select one on the fly. I feel like the actual drawing to a screen surface would never change, but maybe if [it's drawing to a buffer](https://docs.rs/wgpu/latest/wgpu/struct.RenderBundleEncoder.html#method.draw_indirect) you could have different ones for different kinds of effects to create the textures or what-not.

Reply:
I talked a bit with @cwfitzgerald and we decided to just do it for all device resources.

> `BindGroupLayout` is also "yes," I think, because you can share the same bind group layout between pipelines on purpose. It could be useful to check different pipelines' `get_bind_group_layout` return values actually being the same underlying object so you know whether or not you can reuse bind groups between them.

It's worth noting that this isn't and won't be the case, internally we share the same object but at the wgpu level we don't. This is also consistent with the WebGPU backend where you can't tell that 2 bind group layouts point to the same underlying object.

I don't think this is a really big deal, the bind group layouts will still be compatible with each other as per the WebGPU spec.